### ISSUE TITLE
Add an option to export all animations, not just stashed ones. 

### ISSUE THREAD

**Issue by setzer22** [KhronosGroup/glTF-Blender-IO#1258]
**Is your feature request related to a problem? Please describe.**
I have a character model where I'm using the Rigify add-on to generate my rig. Rigify is an auto-rigger, and one of its main features is being able to edit the "meta-rig", even after animations have been created.

However, re-generating the rig in this way has an unintended side effect: All animations are cleared from the NLA strip. Thus, in order to make small adjustments to my rig, as a user, the best workflow I have is:

1. Edit and re-generate the rig
2. Go to the animation tab
3. Select the first action
4. Click on "Stash", to store it on the NLA strip
5. Repeat, until no more actions left.

With a model with 25 actions. An edit as simple as tweaking a bone's position, becomes a huge pain point.

And please, note that this is **not** an issue specific to rigify. I just want to motivate the use case, since I've read a few issues about the topic, and this doesn't seem to be a concern. There's people encountering this issue in other ways that have nothing to do with my workflow (see: https://blender.stackexchange.com/questions/143793/how-to-stash-all-actions-to-current-armature). 

**Describe the solution you'd like**
I would like a way to export my gltf model that is more similar to other export plugins: All actions are exported, not just the ones on the NLA strip.

**Describe alternatives you've considered**
I've been doing this by hand. I could continue doing this, but it's really tedious and I believe could hinder adoption of gltf.
A script like the one suggested in https://blender.stackexchange.com/questions/143793/how-to-stash-all-actions-to-current-armature could also do the trick. However, it doesn't seem to be working anymore, and it's currently beyond my reach to investigate what needs to be done to port this to Blender 2.9. An unofficial, unmaintained script is also not the best idea to solve the issue for people with this problem.

**Additional context**
N/A

**Comment by donmccurdy** [comment]
> ... since I've read a few issues about the topic, and this doesn't seem to be a concern. 

It's certainly something we're aware of, but there's a major obstacle. There is no association between Blender Actions and the objects that they animate, unless the Action is (a) active, (b) stashed, or (c) in an NLA track. When any of those indicators are available, this addon will use the information to export the animation. Without one of the three, we can't. 

If the model has only a single armature, then, yes, it would be possible to assume every animation is intended to go with that armature, but this puts us in a situation where the exporter behaves inconsistently: just adding another object to the scene will break animation export. I'm not sure that inconsistency would be better than the current (sometimes inconvenient) requirement.

**Comment by setzer22** [comment]
@donmccurdy  I understand the issue. Something like this would certainly not work for everyone.

However, it is a very frequent use case for animators to have a single armature in a .blend file representing a single character. That's how it's usually done in videogames, for instance. I believe the large amount of issues in this repo of people confused about their animations not exporting also helps sustain my point.

So I still think there's a lot of value in a feature like this. A configuration flag on export (disabled by default), and a small note on the documentation about animation exporting would really help a lot.

There are two other possible directions for this. One that would be beneficial to everyone (but won't likely happen anytime soon). And the other that would only help my particular use-case (and I'm unsure whether it can even be done):

- Blender could make it easier to stack or add actions to the NLA track. The current UI is a dropdown list, and you must open it, select each animation and click a button. Something like an "Add all" button, or a screen with a list of checkboxes would go a long way in making this way less painful. However, blender development happens slowly, and I'm not even sure if a change like this would be considered or accepted (not likely to fall high on the priority list, to be honest).

- Rigify could avoid clearing the NLA tracks when the rig is regenerated. However, that may not even be a possibility due to the way the rigs are regenerated. It's probably deleting and re-creating the object.

Basically, I feel like on the short-term, this plugin is the best place to add an easy way to solve this issue. After all, for many kinds of animation people don't even use the NLA strip, so this plugin is the only reason I'm aware it exists.

**Comment by chriscamacho** [comment]
I'm honestly at a loss how to export an animation, I have various models that animate in blender, but have nothing in the NLA, I've looked at a number of different tutorials, but I still am no nearer exporting any animation to a gltf file, really doesn't feel like an art path...

could the exporter not have the option to export ALL actions ending in a specific string in the name, other plugins do this for example naming things ending .L or .R

if i could only go into the dope sheet and change the action name ArmatureAction to ArmatureAction.export then in the exporter dialog just add .export into a text field.... optionally if multiple actions could be exported one after the other inside the same animation, this could also make the plugin more usable

**Comment by donmccurdy** [comment]
@chriscamacho for the time being, you need to go into the NLA editor and either *stash* or *push down* each action that you want to export, which associates it with the object that it it supposed to animate. This requires one button click per animation you want to export, and there is no need to rename the animations.

I don't currently know how we can fix this in the addon, without creating even more issues. If someone would like to try writing a PR, that would be welcome, but I would ask that we really need to avoid a situation where every animation gets attached to every object in the scene when there are multiple objects. The fact that Blender Actions do not indicate what they are supposed to animate — unless you use NLA Tracks — makes this tricky.

**Comment by scurest** [comment]
(You should probably file a bug with rigify. If it can fixup the actions, it seems like it should fixup the NLA tracks too. That said....)

**Comment by scurest** [comment]
Here's a go at an implementation: [single-model-multiple-actions](https://github.com/scurest/glTF-Blender-IO/tree/single-model-multiple-actions)

It adds an animation mode drop down in the Animation panel on the export screen.

![fee](https://user-images.githubusercontent.com/11024420/99914308-228ee600-2cc2-11eb-8d75-bdcd14624a91.png)

"Normal" is just like master. "Single Model, Multiple Actions" is new. For it to work, there has to be

* one armature
* actions move pose bones only

It then assumes all actions go on the armature and exports one glTF animation per action.

@setzer22 Can you see if this does what you want?

(There might be bugs, animation data isn't optimized for size, etc.)

**Comment by julienduroure** [comment]
@scurest I worked on another implementation that seems to be simplier (but I didn't tested very well, and with TODO)
https://github.com/KhronosGroup/glTF-Blender-IO/tree/mono_char

**Comment by julienduroure** [comment]
I created a PR #1560 where it exports all animations when blend file contains only 1 armature.
It can lead to exported non wanted actions if some objects in the scene are also animated

**Comment by julienduroure** [comment]
Seems the problem is now solved after merge of #1560. Closing

**Comment by setzer22** [comment]
Glad to see this got finally solved. It's a shame that the better solution (only exporting actions for the selected armatures) is not possible using blender's Python API, but at least now the experience won't be so confusing for first-time users.

I'm sorry I didn't reply to the issue :sweat_smile: I just set out on my madman quest to [build my own 3d modelling software](https://github.com/setzer22/blackjack), so I'm no longer using blender or the gltf export plugin.
